packages/language-service/src/codefixes/fix_missing_import.ts

```diff
@@ -0,0 +1,165 @@
+
+const errorCodes: number[] = [
+  ngErrorCode(NgCompilerErrorCode.SCHEMA_INVALID_ELEMENT),
+];
+
+/**
+ * Updates a given Angular trait, such as an NgModule or standalone Component, by adding
+ * `importName` to the list of imports on the decorator arguments.
+ */
+function updateImportsForAngularTrait(
+    checker: TemplateTypeChecker, trait: ts.ClassDeclaration,
+    importName: string): ts.FileTextChanges|null {
+  // Get the object with arguments passed into the primary Angular decorator for this trait.
+  const decorator = checker.getPrimaryAngularDecorator(trait);
+  if (decorator === null) {
+    return null;
+  }
+  const decoratorProps = findFirstMatchingNode(decorator, {filter: ts.isObjectLiteralExpression});
+  if (decoratorProps === null) {
+    return null;
+  }
+
+  // Update the trait's imports.
+  const newDecoratorProps =
+      updateObjectValueForKey(decoratorProps, 'imports', (oldValue?: ts.Expression) => {
+        if (oldValue && !ts.isArrayLiteralExpression(oldValue)) {
+          return oldValue;
+        }
+        return ensureArrayWithIdentifier(ts.factory.createIdentifier(importName), oldValue);
+      });
+
+  return {
+    fileName: trait.getSourceFile().fileName,
+    textChanges: [{
+      span: {
+        start: decoratorProps.getStart(),
+        length: decoratorProps.getEnd() - decoratorProps.getStart()
+      },
+      newText: printNode(newDecoratorProps, trait.getSourceFile())
+    }],
+  };
+}
+
+/**
+ * This code action will generate a new import for an unknown selector.
+ */
+export const missingImportMeta: CodeActionMeta = {
+  errorCodes,
+  getCodeActions: (
+      {templateInfo, start, compiler, formatOptions, preferences, errorCode, tsLs}) => {
+    const checker = compiler.getTemplateTypeChecker();
+
+    // The error must be an invalid element in tag, which is interpreted as an intended selector.
+    const target = getTargetAtPosition(templateInfo.template, start);
+    if (target === null || target?.context.kind !== TargetNodeKind.ElementInTagContext ||
+        target.context.node instanceof t.Template) {
+      return [];
+    }
+    const missingElement = target.context.node;
+    const missingElementTag = target.context.node.name;
+
+    // The class which has an imports array; either a standalone trait or its owning NgModule.
+    const importOn = checker.getOwningNgModule(templateInfo.component) ?? templateInfo.component;
+
+    // Find all possible importable directives with a matching selector, and take one of them.
+    // In the future, we could handle multiple matches as additional quick fixes.
+    const allPossibleDirectives = checker.getPotentialTemplateDirectives(templateInfo.component);
+    const matchingDirectives =
+        getDirectiveMatchesForElementTag(missingElement, allPossibleDirectives);
+    if (matchingDirectives.size === 0) {
+      return [];
+    }
+    const bestMatch = [...matchingDirectives][0];
```

Comment:
```suggestion
    const bestMatch = matchingDirectives[0];
```
Not sure if there was a reason for the array copy here? I'm guessing there was a refactor at some point

Reply:
Indeed.